Storage/tests/System/KmsTest.php

```diff
@@ -0,0 +1,283 @@
+    {
+        $keyNames = [];
+        $keyFilePath = getenv('GOOGLE_CLOUD_PHP_TESTS_KEY_PATH');
+        $wrapper = new RequestWrapper([
+            'keyFile' => json_decode(file_get_contents($keyFilePath), true),
+            'scopes' => ['https://www.googleapis.com/auth/cloud-platform']
+        ]);
+        $projectId = self::getProjectId($keyFilePath);
+
+        self::buildKeyRing($wrapper, $projectId, $keyRingId);
+        $keyNames[] = self::getCryptoKeyName($wrapper, $projectId, $keyRingId, $cryptoKeyId1);
+        $keyNames[] = self::getCryptoKeyName($wrapper, $projectId, $keyRingId, $cryptoKeyId2);
+
+        return $keyNames;
+    }
+
+    /**
+     * @param RequestWrapper $wrapper
+     * @param string $projectId
+     * @param string $keyRingId
+     */
+    private static function buildKeyRing(
+        RequestWrapper $wrapper,
+        $projectId,
+        $keyRingId
+    ) {
+        try {
+            $wrapper->send(
+                new Request(
+                    'POST',
+                    sprintf(
+                        'https://cloudkms.googleapis.com/v1/projects/%s/locations/global/keyRings?keyRingId=%s',
+                        $projectId,
+                        $keyRingId
+                    )
+                )
+            );
+        } catch (ConflictException $ex) {
+            // If it already exists, great!
+        }
+    }
+
+    /**
+     * @param RequestWrapper $wrapper
+     * @param string $projectId
+     * @param string $keyRingId
+     * @param string $cryptoKeyId
+     * @return string
+     */
+    private static function getCryptoKeyName(
+        RequestWrapper $wrapper,
+        $projectId,
+        $keyRingId,
+        $cryptoKeyId
+    ) {
+        $name = null;
+
+        try {
+            $response = $wrapper->send(
+                new Request(
+                    'POST',
+                    sprintf(
+                        'https://cloudkms.googleapis.com/v1/projects/%s/locations/global/keyRings/%s/cryptoKeys?cryptoKeyId=%s',
+                        $projectId,
+                        $keyRingId,
+                        $cryptoKeyId
+                    ),
+                    [],
+                    json_encode(['purpose' => 'ENCRYPT_DECRYPT'])
+                )
+            );
+
+            $name = json_decode((string) $response->getBody(), true)['name'];
+        } catch (ConflictException $ex) {
+            $name = sprintf(
+                'projects/%s/locations/global/keyRings/%s/cryptoKeys/%s',
+                $projectId,
+                $keyRingId,
+                $cryptoKeyId
+            );
+        }
+
+        $policy = [
+            'policy' => [
+                'bindings' => [
+                    [
+                        'role' => 'roles/cloudkms.cryptoKeyEncrypterDecrypter',
+                        'members' => [
+                            "serviceAccount:$projectId@gs-project-accounts.iam.gserviceaccount.com"
```

Comment:
I recommend to use instead of writing the form `project-id@gs-project-accounts.iam.gserviceaccount.com` as this will change to `service-project-number@gs-project-accounts.iam.gserviceaccount.com` for example `service-00877902500@gs-project-accounts.iam.gserviceaccount.com`.
https://cloud.google.com/storage/docs/json_api/v1/projects/serviceAccount/get

Reply:
Ah, okay. We pull the projectId from the credentials file. Will newly regenerated credentials have the project number instead now, or will we need to trigger a network request to get it?

Reply:
The new format will be provided by the serviceAccount.get request. It's a robot service account used by GCS to get permission to use KMS or other GCP services.

Clarifying note: My recommendation is to request the robot service account using serviceAccount.get to make sure you're using the right one. The format will be changed later this quarter so you'll still see the old project-id@gs-project... format right now.

Reply:
Sounds good! I will update to pull this data using the GET request.